device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java

```diff
@@ -1336,6 +1022,207 @@ public void onSubscriptionMessageSent(int deliveryTag, SubscriptionType subscrip
         this.inProgressSubscriptionMessages.put(deliveryTag, subscriptionType);
     }
 
+    @Override
+    public void onDeviceSessionOpened(String deviceId)
+    {
+        this.deviceSessionsOpenedLatch.countDown();
+    }
+
+    @Override
+    public void onAuthenticationSessionOpened() {
+        this.authenticationSessionOpenedLatch.countDown();
+    }
+
+    final void addDeviceSession(DeviceClientConfig deviceClientConfig, SubscriptionMessageRequestSentCallback subscriptionMessageRequestSentCallback, boolean afterOpen) throws TransportException {
+        if (deviceClientConfig == null)
+        {
+            throw new IllegalArgumentException("deviceClientConfig cannot be null.");
+        }
+
+        // Check if the device session still exists from a previous connection
```

Comment:
What does "previous connection" mean here?

Reply:
In the IotHubTransport layer, there is some retry logic that recycles the instance of AmqpsIotHubConnection.java between retry attempts. As such, this layer was written with the understanding that there may be states (like device foo being subscribed to twin) that should carry over from one connection to another